elbepack/initvmaction.py

```diff
@@ -269,7 +392,33 @@ class AttachAction(InitVMAction):
     def __init__(self, node):
         InitVMAction.__init__(self, node)
 
-    def execute(self, _initvmdir, _opt, _args):
+    def _attach_qemu_vm(self, initvmdir):
+        """
+        Attach to QEMU initvm.
+
+        This method is using socat to connect to the unix-socket of the 
+        serial console of the initvm.
+        """
+        initvmdir = check_initvm_dir(initvmdir)
+
+        # Test if socat command is available.
+        try:
+            subprocess.run('which socat > /dev/null 2>&1', shell=True, check=True)
+        except Exception:
+            print('The command "socat" is required.\nPlease install socat: "sudo apt install socat"')
+            sys.exit(208)
+        
+        # Connect to socket file, if it exists.
+        if os.path.exists(os.path.join(initvmdir, 'vm-serial-socket')):
+            subprocess.run(f'cd {initvmdir} && socat stdin,raw,echo=0,escape=0x11 unix-connect:vm-serial-socket',
```

Comment:
You should be able to replace this with something like (fairly untested):
```py
selector = selectors.DefaultSelector()
stdin_rdy = selector.register(sys.stdin.buffer, selectors.EVENT_READ)
serial_rdy = selector.register(serial_socket, selectors.EVENT_READ)

while True:
    ready = selectors.select()
    for key, events in ready:
        if key == stdin_rdy:
            serial_rdy.send(stdin.buffer.read())
        elif key == serial_rdy:
            sys.stdin.buffer.write(serial_socket.read())
```

Reply:
Which in fact is implementing a shell client.
The old libvirt implementation has for the attach command has a similar implementation as my implementation for QEMU, and I don't see it as part of this PR to implement a shell client in Python.